text/utilities/0003-git-credential-configuration.md

```diff
@@ -0,0 +1,59 @@
+# Credential Configuration in the Git Buildpack
+
+## Summary
+
+Add the ability to supply credentials to the Git buildpack that will be used by
+`git` in order to allow later buildpacks to perform operations that require
+`git` authentication.
+
+## Motivation
+
+There are some operations, such as obtaining private language modules, that
+require `git` authentication during the build process. Currently there is no
+way of authenticating as a user to enable these workflows. Adding this
+functionality to the Git buildpack would open up a wider range of workflows
+than adding the functionality on an individual buildpack basis and make the
+ability to add credentials to the project universally.
+
+## Implementation
+
+For security purposes, the credentials cannot be provided through an
+environment variable like most of our other buildpack configuration as this
+poses a risk of exposure by view the build process in some form of CI. Because
+of this restriction, an approach similar to the CA Certificates Buildpack will
+be used where the credentials will be provided in a mounted volume through a
+service binding. Those credentials will then be used to change the global `git`
+config for that container which will persist for the entirety of the build
+which the [CNB specification](https://github.com/buildpacks/spec/blob/main/buildpack.md#requirements)
+states must all happen in the same container for security purposes.
+
+The service binding structure will be as follows:
+- `type`: `git-credentials`
+- `context` (optional): The context is an optional pattern as defined by
+  [`git`](https://git-scm.com/docs/gitcredentials#_credential_contexts). If a
+  context is not provided then the credentials given in the binding will be the
+  default credentials the `git` uses when authenticating. A given context can
+  only be used once for any group of bindings, if a context is given by two
+  separate bindings the build will fail.
+- `credentials`: The credentials file should have the following format to
+  conform with the [`git` credential structure](https://git-scm.com/docs/git-credential#IOFMT).
+```
+username=some-username
+password=some-password/token
+```
+
+For every service binding the following `git` command will be run:
+```shell
+git config --global credential.{{context}}.helper "!cat path/to/binding/credentials"
```

Comment:
Does the tiny builder have a shell in it? or does git run this command in a way that does not require a shell?

Reply:
The build image for tiny does have a shell. These settings will only be present for build, I did not see much purpose in configuring this to run during launch as well as it feels mostly pertinent to building. Are there applications that would require git credentials during run?

Reply:
That was my assumption too. That we'd only be providing this support for build time.

I guess it's possible that an application would need to use git, but I'd say it's up to the application to configure its credentials. A good 12-factor app shouldn't depend on that being pre-configured at the OS.

My $0.02 is that it's OK to only focus on build so long as we document that.